File: app/src/main/java/com/twr/mangago/MainActivity.java
package com.twr.mangago;


import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.core.view.WindowInsetsCompat;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import android.os.Build;
import android.os.Bundle;
import android.util.Base64;
import android.util.Log;
import android.view.GestureDetector;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowInsets;
import android.view.WindowInsetsController;
import android.webkit.ValueCallback;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.RelativeLayout;


import androidx.appcompat.widget.Toolbar;

import com.google.android.material.color.DynamicColors;

import java.io.InputStream;



public class MainActivity extends AppCompatActivity {
    private WebView webView;
    private SwipeRefreshLayout swipeRefresh;
    private RelativeLayout mainContainer;
    Boolean multipage = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        DynamicColors.applyToActivityIfAvailable(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        webView = findViewById(R.id.webview);
        Toolbar toolBar = findViewById(R.id.toolBar);
        setSupportActionBar(toolBar);
        swipeRefresh = findViewById(R.id.swipeContainer);
        swipeRefresh.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                webView.reload();
            }
        });
        webView.setOnTouchListener(new View.OnTouchListener() {
            GestureDetector gestureDetector = new GestureDetector(getApplicationContext(), new GestureDetector.SimpleOnGestureListener() {
                @Override
                public boolean onSingleTapUp(MotionEvent e) {
                    /* if (toolBar.getVisibility() == View.GONE ){
                        toolBar.setVisibility(View.VISIBLE);
                    }
                    else{
                        toolBar.setVisibility(View.GONE);
                    } */
                return super.onSingleTapUp(e);
                }
            });

            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                /* if (checkUrl(webView.getUrl()))
                {
                gestureDetector.onTouchEvent(motionEvent);} */
                return false;
            }
        });
        loadWeb();
    }

    public void loadWeb() {
        mainContainer = findViewById(R.id.mainContainer);
        webView = findViewById(R.id.webview);
        webView.getSettings().setDomStorageEnabled(true);
        webView.loadUrl("https://www.mangago.me/");
        webView.getSettings().setJavaScriptEnabled(true);
        webView.getSettings().setBuiltInZoomControls(true);
        webView.getSettings().setDisplayZoomControls(false);
        webView.setOverScrollMode(View.OVER_SCROLL_NEVER);
        swipeRefresh.setRefreshing(true);
        webView.setWebViewClient(new WebViewClient() {
                 @Override
                 public void onPageFinished(WebView view, String url) {
                     injectCSS();
                     super.onPageFinished(view, url);
                     swipeRefresh.setRefreshing(false);
                     showSystemBars();
                     checkUrl(url);
                 }
             }
        );
    }

    private void checkUrl(String url) {
        webView.evaluateJavascript("(function() { var element = document.getElementById('reader-nav'); return element.innerHTML; })();",
                s -> {
                    if ( s.equals("null") || s.equals("undefined")) {
                        Log.v("checkURL", s);
                        showSystemBars();
                        mainContainer.setFitsSystemWindows(true);
                    } else {
                        Log.v("checkURL", s);
                        hideSystemBars();
                        mainContainer.setFitsSystemWindows(false);
                        if (!multipage) {
                            // Force multi-page to be enabled by default
                            webView.evaluateJavascript("document.getElementById('multi_page').checked = false;" +
                                    "document.getElementById('multi_page_form').submit();", null);
                            multipage = true;
                        }
                    }
                });
    }

    private void injectCSS() {
        try {
            InputStream inputStream = getAssets().open("style.css");
            byte[] buffer = new byte[inputStream.available()];
            inputStream.read(buffer);
            inputStream.close();
            String encoded = Base64.encodeToString(buffer, Base64.NO_WRAP);
            webView.loadUrl("javascript:(function() {" +
                    "var parent = document.getElementsByTagName('head').item(0);" +
                    "var style = document.createElement('style');" +
                    "style.type = 'text/css';" +
                    // Tell the browser to BASE64-decode the string into your script !!!
                    "style.innerHTML = window.atob('" + encoded + "');" +
                    "parent.appendChild(style)" +
                    "})()");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void hideSystemBars() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            getWindow().setDecorFitsSystemWindows(false);
            WindowInsetsController controller = getWindow().getInsetsController();
            if (controller != null) {
                controller.hide(WindowInsets.Type.statusBars() | WindowInsets.Type.navigationBars());
                controller.setSystemBarsBehavior(WindowInsetsController.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE);
            }
            getSupportActionBar().hide();
        } else {
            getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    | View.SYSTEM_UI_FLAG_FULLSCREEN
                    | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                    | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                    | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                    | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
            );
        }

    }

    private void showSystemBars() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            getWindow().setDecorFitsSystemWindows(true);
            WindowInsetsController controller = getWindow().getInsetsController();
            if (controller != null) {
                controller.show(WindowInsetsCompat.Type.systemBars());
            }
            getSupportActionBar().show();
        } else {
            getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    | View.SYSTEM_UI_FLAG_FULLSCREEN
                    | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                    | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                    | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                    | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY);
        }
    }

    @Override
    public void onBackPressed() {

        if (webView.isFocused() && webView.canGoBack()) {
            webView.goBack();
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(@NonNull Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu, menu);
        MenuItem menuItem = menu.findItem(R.id.search);
        SearchView searchView = (SearchView) menuItem.getActionView();
        searchView.setQueryHint("Search mangas");
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                webView.loadUrl("https://www.mangago.me/r/l_search/?name=" + query);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });
        return true;
    }


    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.register:
                webView.loadUrl("https://www.mangago.me/home/accounts/register/");
                return true;
            case R.id.signin:
                webView.loadUrl("https://www.mangago.me/home/accounts/login/");
                return true;
            case R.id.home:
                webView.loadUrl("https://www.mangago.me/");
                return true;
            case R.id.history:
                webView.loadUrl("https://www.mangago.me/home/history/");
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

}

File: app/src/main/java/com/recreation/mangago/MainActivity.kt
package com.recreation.mangago

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Bitmap
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Base64
import android.util.Log
import android.view.*
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.RelativeLayout
import androidx.appcompat.widget.SearchView
import androidx.core.view.WindowInsetsCompat
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.color.DynamicColors
import java.lang.Exception

class MainActivity : AppCompatActivity() {
    private val webView: WebView by lazy { findViewById(R.id.webView) }
    val swipeRefresh: SwipeRefreshLayout by lazy { findViewById(R.id.swipeContainer) }
    val mainContainer: RelativeLayout by lazy { findViewById(R.id.mainContainer) }

    override fun onCreate(savedInstanceState: Bundle?) {
        DynamicColors.applyToActivityIfAvailable(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(findViewById(R.id.toolBar))
        swipeRefresh.setOnRefreshListener { webView.reload() }
        loadWeb()
    }

    @SuppressLint("SetJavaScriptEnabled")
    fun loadWeb() {
        val webViewSettings = webView.settings
        webViewSettings.domStorageEnabled = true
        webViewSettings.javaScriptEnabled = true
        webViewSettings.builtInZoomControls = true
        webViewSettings.displayZoomControls = false
        webView.overScrollMode = View.OVER_SCROLL_NEVER
        webView.webViewClient = CustomWebViewClient()
        webView.loadUrl("https://www.mangago.me/")
    }

    inner class CustomWebViewClient : WebViewClient() {
        override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
            swipeRefresh.isRefreshing = true
            super.onPageStarted(view, url, favicon)
        }

        override fun onPageFinished(view: WebView?, url: String?) {
            injectCSS()
            super.onPageFinished(view, url)
            swipeRefresh.isRefreshing = false
            checkURL(url!!)
        }

        private fun injectCSS() {
            try {
                val inputStream = assets.open("style.css")
                val buffer = ByteArray(inputStream.available())
                inputStream.read(buffer)
                inputStream.close()
                val encoded = Base64.encodeToString(buffer, Base64.NO_WRAP)
                webView.loadUrl(
                    "javascript:(function() {" +
                            "var parent = document.getElementsByTagName('head').item(0);" +
                            "var style = document.createElement('style');" +
                            "style.type = 'text/css';" +
                            "style.innerHTML = window.atob('" + encoded + "');" +
                            "parent.appendChild(style)" +
                            "})()"
                )
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

    }

    private fun checkURL(url: String) {
        var count = 0
        for (char in url) {
            if (char == '/') {
                count++
            }
        }
        if (webView.progress == 100) {
            if (count == 8) {
                Log.v("READER", "COUNT")
                startReader(url)
            } else {
                webView.evaluateJavascript(
                    "(function() { var element = document.getElementById('reader-nav'); return element.innerHTML; })();"
                ) { s: String ->
                    if (s == "null" || s == "undefined") {
                        Log.v("NULL", "NULL")
                    } else {
                        startReader(url)
                        Log.v("READER", "JS")
                    }
                }
            }
        }
    }

    private fun startReader(url: String) {
        webView.goBack()
        val intent = Intent(this, MangaReaderActivity::class.java)
        intent.putExtra("MESSAGE", url)
        startActivity(intent)
    }

    override fun onBackPressed() {
        if (webView.isFocused && webView.canGoBack()) {
            webView.goBack()
        } else {
            super.onBackPressed()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu, menu)
        val menuItem = menu.findItem(R.id.search)
        val searchView = menuItem.actionView as SearchView
        searchView.queryHint = "Search mangas"
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                webView.loadUrl("https://www.mangago.me/r/l_search/?name=$query")
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                return false
            }
        })
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.register -> {
                webView.loadUrl("https://www.mangago.me/home/accounts/register/")
                return true
            }
            R.id.signin -> {
                webView.loadUrl("https://www.mangago.me/home/accounts/login/")
                return true
            }
            R.id.home -> {
                webView.loadUrl("https://www.mangago.me/")
                return true
            }
            R.id.history -> {
                webView.loadUrl("https://www.mangago.me/home/history/")
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }
}